File: src/carrot_switch/config/claude.py
"""Claude Code configuration CRUD.

Claude Code stores MCP servers in ~/.claude.json under the "mcpServers" key.
Skills are stored in ~/.claude/skills/.

Key differences from OpenCode/MiMoCode:
- command is a string (not array)
- env (not environment)
- No 'enabled' field (all servers are active)
- type: "http" for remote (not "remote")
"""
from pathlib import Path
from carrot_switch.config import read_jsonc, write_jsonc


def _get_config_path() -> Path:
    return Path.home() / ".claude.json"


def get_config_path() -> Path:
    return _get_config_path()


def get_skills_dir() -> Path:
    return Path.home() / ".claude" / "skills"


def is_available() -> bool:
    return _get_config_path().exists()


def get_mcp_servers() -> dict:
    """Read MCP servers from Claude config.

    Returns normalized format (array command, environment key) for internal use.
    """
    cfg = read_jsonc(_get_config_path())
    raw_servers = cfg.get("mcpServers", {})
    normalized = {}
    for name, server in raw_servers.items():
        normalized[name] = _normalize_from_claude(server)
    return normalized


def _normalize_from_claude(server: dict) -> dict:
    """Convert Claude format to internal format."""
    result = dict(server)

    # Add default type if missing (Claude doesn't have a type field for local servers)
    if "type" not in result:
        result["type"] = "local"

    # Convert command string to array
    if isinstance(result.get("command"), str):
        cmd = result.pop("command")
        args = result.pop("args", [])
        result["command"] = [cmd] + args

    # Rename env to environment
    if "env" in result:
        result["environment"] = result.pop("env")

    # Map type: "http" to "remote"
    if result.get("type") == "http":
        result["type"] = "remote"

    # Add enabled field (Claude doesn't have it)
    if "enabled" not in result:
        result["enabled"] = True

    return result


def _to_claude_format(server: dict) -> dict:
    """Convert internal format to Claude format."""
    result = dict(server)

    # Convert command array to string + args
    cmd_list = result.pop("command", [])
    if cmd_list:
        result["command"] = cmd_list[0]
        if len(cmd_list) > 1:
            result["args"] = cmd_list[1:]

    # Rename environment to env
    if "environment" in result:
        result["env"] = result.pop("environment")

    # Map type: "remote" to "http"
    if result.get("type") == "remote":
        result["type"] = "http"

    # Remove enabled (Claude doesn't use it)
    result.pop("enabled", None)

    return result


def add_mcp_server(name: str, server: dict) -> None:
    cfg = read_jsonc(_get_config_path())
    claude_server = _to_claude_format(server)
    cfg.setdefault("mcpServers", {})[name] = claude_server
    write_jsonc(_get_config_path(), cfg)


def update_mcp_server(name: str, server: dict) -> None:
    cfg = read_jsonc(_get_config_path())
    if name not in cfg.get("mcpServers", {}):
        raise KeyError(f"MCP server '{name}' not found")
    claude_server = _to_claude_format(server)
    cfg["mcpServers"][name] = claude_server
    write_jsonc(_get_config_path(), cfg)


def delete_mcp_server(name: str) -> None:
    cfg = read_jsonc(_get_config_path())
    servers = cfg.get("mcpServers", {})
    if name not in servers:
        raise KeyError(f"MCP server '{name}' not found")
    del servers[name]
    write_jsonc(_get_config_path(), cfg)


def get_skills_permission() -> dict:
    """Claude doesn't have a permission system for skills like OpenCode/MiMoCode."""
    return {}


def set_skill_permission(name: str, allowed: bool) -> None:
    """Claude doesn't have a permission system for skills."""
    pass

File: server/src/lib/config/claude.ts
import { join } from "path";
import { homedir } from "os";
import { existsSync } from "fs";
import { readJsonc, writeJsonc } from "../jsonc.js";

let _configPathOverride: string | null = null;

export function _setConfigPathForTesting(path: string | null) {
  _configPathOverride = path;
}

function getConfigPath(): string {
  if (_configPathOverride) return _configPathOverride;
  return join(homedir(), ".claude.json");
}

function getSkillsDirPath(): string {
  return join(homedir(), ".claude", "skills");
}

export function get_config_path(): string {
  return getConfigPath();
}

export function get_skills_dir(): string {
  return getSkillsDirPath();
}

export function is_available(): boolean {
  return existsSync(getConfigPath());
}

function normalizeFromClaude(server: Record<string, any>): Record<string, any> {
  const result = { ...server };

  if (!result.type) result.type = "local";

  if (typeof result.command === "string") {
    const cmd = result.command;
    const args = result.args || [];
    result.command = [cmd, ...args];
    delete result.args;
  }

  if (result.env) {
    result.environment = result.env;
    delete result.env;
  }

  if (result.type === "http") result.type = "remote";

  if (result.enabled === undefined) result.enabled = true;

  return result;
}

function toClaudeFormat(server: Record<string, any>): Record<string, any> {
  const result = { ...server };

  const cmdList = result.command || [];
  delete result.command;
  if (cmdList.length > 0) {
    result.command = cmdList[0];
    if (cmdList.length > 1) {
      result.args = cmdList.slice(1);
    }
  }

  if (result.environment) {
    result.env = result.environment;
    delete result.environment;
  }

  if (result.type === "remote") result.type = "http";

  delete result.enabled;

  return result;
}

export function get_mcp_servers(): Record<string, any> {
  const cfg = readJsonc(getConfigPath());
  const rawServers = cfg.mcpServers || {};
  const normalized: Record<string, any> = {};
  for (const [name, server] of Object.entries(rawServers)) {
    normalized[name] = normalizeFromClaude(server as Record<string, any>);
  }
  return normalized;
}

export function add_mcp_server(name: string, server: Record<string, any>): void {
  const configPath = getConfigPath();
  const cfg = readJsonc(configPath);
  if (!cfg.mcpServers) cfg.mcpServers = {};
  cfg.mcpServers[name] = toClaudeFormat(server);
  writeJsonc(configPath, cfg);
}

export function update_mcp_server(name: string, server: Record<string, any>): void {
  const configPath = getConfigPath();
  const cfg = readJsonc(configPath);
  if (!cfg.mcpServers || !(name in cfg.mcpServers)) {
    throw new Error(`MCP server '${name}' not found`);
  }
  cfg.mcpServers[name] = toClaudeFormat(server);
  writeJsonc(configPath, cfg);
}

export function delete_mcp_server(name: string): void {
  const configPath = getConfigPath();
  const cfg = readJsonc(configPath);
  const servers = cfg.mcpServers || {};
  if (!(name in servers)) {
    throw new Error(`MCP server '${name}' not found`);
  }
  delete servers[name];
  writeJsonc(configPath, cfg);
}

export function get_skills_permission(): Record<string, string> {
  return {};
}

export function set_skill_permission(_name: string, _allowed: boolean): void {
  // Claude doesn't have a permission system for skills
}
